pkg/controllers/nodeBackend/nodeBackend.go

```diff
@@ -0,0 +1,496 @@
+		for _, taint := range node.Spec.Taints {
+			if taint.Key == ToBeDeletedTaint {
+				klog.V(4).Infof("Ignoring node %v with autoscaler taint %+v", node.Name, taint)
+				return false
+			}
+		}
+
+		// If we have no info, don't accept
+		if len(node.Status.Conditions) == 0 {
+			return false
+		}
+		for _, cond := range node.Status.Conditions {
+			// We consider the node for load balancing only when its NodeReady condition status
+			// is ConditionTrue
+			if cond.Type == corev1.NodeReady && cond.Status != corev1.ConditionTrue {
+				klog.V(4).Infof("Ignoring node %v with %v condition status %v", node.Name, cond.Type, cond.Status)
+				return false
+			}
+		}
+		return true
+	}
+}
+
+func (c *Controller) ensureBackends(ingressClass *networkingv1.IngressClass, lbID string) error {
+
+	ingresses, err := c.getIngressesForClass(ingressClass)
+	if err != nil {
+		return err
+	}
+	nodes, err := listWithPredicate(c.nodeLister, c.getNodeConditionPredicate())
+	if err != nil {
+		return err
+	}
+	// If there are no available nodes.
+	if len(nodes) == 0 {
+		return fmt.Errorf("error: %s due to : %s", "UnAvailableNodes", "There are no available provisioned nodes")
+	}
+
+	for _, ingress := range ingresses {
+		for _, rule := range ingress.Spec.Rules {
+			for _, path := range rule.HTTP.Paths {
+				svcName, svcPort, _, svc, err := c.pathToServiceAndTargetPort(ingress.Namespace, path)
+				if err != nil {
+					return err
+				}
+
+				if svc == nil || svc.Spec.Ports == nil || svc.Spec.Ports[0].NodePort == 0 {
+					continue
+				}
+
+				pods, err := c.retrievePods(err, ingress, svcName, svcPort)
+				if err != nil {
+					return err
+				}
+
+				var backends []ociloadbalancer.BackendDetails
+				nodePort := svc.Spec.Ports[0].NodePort
+				trafficPolicy := svc.Spec.ExternalTrafficPolicy
+				if trafficPolicy == corev1.ServiceExternalTrafficPolicyTypeCluster {
+					for _, node := range nodes {
+						backends = append(backends, newBackend(NodeInternalIP(node), nodePort))
+					}
+				} else {
+					for _, pod := range pods {
+						node, err := c.nodeLister.Get(pod.Spec.NodeName)
+						if err != nil {
+							if apierrors.IsNotFound(err) {
+								klog.Infof("node %s has been deleted, skipping pod", pod.Spec.NodeName)
+								continue
+							}
+							return err
+						}
+						backends = append(backends, newBackend(NodeInternalIP(node), nodePort))
+					}
+				}
+				backendSetName := util.GenerateBackendSetName(ingress.Namespace, svcName, svcPort)
+				err = c.client.GetLbClient().UpdateBackends(context.TODO(), lbID, backendSetName, backends)
+				if err != nil {
+					return fmt.Errorf("unable to update backends for %s/%s: %w", ingressClass.Name, backendSetName, err)
+				}
+			}
+		}
+	}
+	// Sync default backends
+	c.syncDefaultBackend(lbID, ingresses)
+	return nil
+}
+
+func (c *Controller) retrievePods(err error, ingress *networkingv1.Ingress, svcName string, svcPort int32) ([]*corev1.Pod, error) {
```

Comment:
Refactor pod retrieval to a generic place.

Reply:
Done